foundation/org.eclipse.persistence.core/src/main/java/org/eclipse/persistence/asm/Opcodes.java

```diff
@@ -22,6 +22,7 @@
 
 public class Opcodes {
 
+    //This block must be first - begin
```

Comment:
To avoid a static initialization order problem with ASMFactory, add the following two lines here:

    public final static String ASM_SERVICE_ECLIPSELINK = "eclipselink";
    public final static String ASM_SERVICE_OW2 = "ow2";
    
    Then change lines 32 and 33 below to use the above variables:
    
        ASM_OPCCODES_MAP.put(ASM_SERVICE_OW2, ASM_OPCCODES_OW2);
        ASM_OPCCODES_MAP.put(ASM_SERVICE_ECLIPSELINK, ASM_OPCCODES_ECLIPSELINK);

Reply:
@anija-anil My prior recommendations do not resolve the issue because Opcodes needs to reference ASMFactory in order to initialize all the static fields.  Looks like the only way to resolve  the circular initialization is to change ASMFactory to not reference any static fields in Opcodes.... so disregard the prior suggestion and instead do something like this in ASMFactory:

```
    private final static String ASM_ECLIPSELINK_CLASS_VISITOR = "org.eclipse.persistence.internal.libraries.asm.ClassVisitor";

    // Note: Do not reference static fields in Opcodes to avoid circular static initialization

    //Should be changed in case of ASM upgrade
    public final static int ASM_API_SELECTED = valueInt("ASM9");
    public final static int JAVA_CLASS_VERSION = valueInt("V1_8");

...
    private static int valueInt(String fieldName) {
        return ((int) Util.getFieldValue(ASM_OPCCODES_MAP, fieldName, Integer.TYPE));
    }
```
This approach eliminates all dependencies that ASMFactory has on Opcodes static fields.

Reply:
Resoloved